### PR TITLE
[resource manager]: add separated resouce manager interface

### LINKED ISSUE #16: Resource allocation should be implemented by the ResourceManager
https://github.com/hanfei1991/mircocosom/blob/f9f0de8e04d79c1fe2c3567c24b5b80c4524f889/master/jobmaster/benchmark/master.go#L122

JobMaster focus on how to create subjob and monitor task status etc.

### PR DESCRIPTION
**Pull request by amyangfei**
A refactor for #4, also close #16

- Add a standalone resource package, and a `RescMgr` interface.
- The executor manager interacts with `RescMgr` as follows
   - The executor manager calls `Register` and `Update` API provided by `RescMgr` when new executor is registered, or status/usage update.
   - The executor manager asks `RescMgr` for resource
   - The `RescMgr` allocates executors to tasks

After this refactor, server master doesn't need to interact with resource directly, instead it works as follows

server master <-> executor manager <-> resource manager

### CODE REVIEW

File: master/resource/manager.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package resource
+
+import (
+	"github.com/hanfei1991/microcosm/model"
+	"github.com/hanfei1991/microcosm/pb"
+)
+
+// RescMgrr manages the resources of the clusters.
```
Comment (amyangfei) [inline master/resource/manager.go:8]:
```suggestion
// RescMgr manages the resources of the clusters.
```

File: master/resource/capacity_impl.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+import (
+	"sync"
+
+	"github.com/hanfei1991/microcosm/model"
+	"github.com/hanfei1991/microcosm/pb"
+	"github.com/hanfei1991/microcosm/pkg/errors"
+)
+
+// CapRescMgr implements ResourceMgr interface, and it uses node capacitiy as
```
Comment (hanfei1991) [inline master/resource/capacity_impl.go:11]:
```suggestion
// CapRescMgr implements ResourceMgr interface, and it uses node capacity as
```

File: master/resource/capacity_impl.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	}
+}
+
+// Allocate implements RescMgr.Allocate
+func (m *CapRescMgr) Allocate(tasks []*pb.ScheduleTask) (bool, *pb.TaskSchedulerResponse) {
+	return m.allocateTasksWithNaiveStrategy(tasks)
+}
+
+// Update implements RescMgr.Update
+func (m *CapRescMgr) Update(id model.ExecutorID, use RescUnit, status model.ExecutorStatus) error {
```
Comment (hanfei1991) [inline master/resource/capacity_impl.go:40]: When an executor is put offline, we should also remove it from resource manager.

File: master/resource/capacity_impl.go (hunk 2 above)
Comment (amyangfei) [reply master/resource/capacity_impl.go:11]: addressed

File: master/resource/capacity_impl.go (hunk 3 above)
Comment (amyangfei) [reply master/resource/capacity_impl.go:40]: Good catch, addressed

File: master/jobmaster/system/impl.go [hunk 4]
```diff
@@ -22,9 +22,8 @@ type Master struct {
 	ctx    context.Context
 	cancel func()
 
-	resourceManager cluster.ResourceMgr
-	client          cluster.ExecutorClient
-	mClient         cluster.JobMasterClient
+	client  cluster.ExecutorClient
+	mClient cluster.JobMasterClient
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply master/jobmaster/system/impl.go:26]: This change is caused by gofmt, and is not relevant to this PR. I will refactor job master in another PR.

**Comment by amyangfei** [comment]
PTAL @hanfei1991

**Review by hanfei1991 — APPROVED** [review]

**Pull request merged**